File: core/src/main/java/net/rickiekarp/core/ui/windowmanager/WindowStack.java
package net.rickiekarp.core.ui.windowmanager;

import net.rickiekarp.core.view.MainScene;
import javafx.scene.Node;
import javafx.scene.control.Button;

import java.util.Stack;

public class WindowStack extends Stack<Node> {
    private Button backButton;

    @Override
    public Node push(Node item) {
        //System.out.println("Push " + item);
        addElement(item);
        MainScene.mainScene.getBorderPane().setCenter(item);
        if (this.size() == 2) {
            initializeBackButton();
            MainScene.mainScene.getWindowScene().getWin().getTitlebarButtonBox().getChildren().add(backButton);
        }
        return item;
    }

    @Override
    public Node pop() {
        //System.out.print("pop -> " + this.peek());
        return super.pop();
    }

    private void initializeBackButton() {
        backButton = new Button("Back");
        backButton.setOnAction(event -> {
            this.pop();
            MainScene.mainScene.getBorderPane().setCenter(super.peek());
            if (this.size() == 1) {
                MainScene.mainScene.getWindowScene().getWin().getTitlebarButtonBox().getChildren().remove(backButton);
                backButton = null;
            }
        });
    }
}

File: projects/JavaFXApps/core/src/main/java/net/rickiekarp/core/ui/windowmanager/WindowStack.kt
package net.rickiekarp.core.ui.windowmanager

import net.rickiekarp.core.view.MainScene
import javafx.scene.Node
import javafx.scene.control.Button

import java.util.Stack

class WindowStack : Stack<Node>() {
    private var backButton: Button? = null

    override fun push(item: Node): Node {
        //System.out.println("Push " + item);
        addElement(item)
        MainScene.mainScene.borderPane.center = item
        if (this.size == 2) {
            initializeBackButton()
            MainScene.mainScene.windowScene!!.win.titlebarButtonBox.children.add(backButton)
        }
        return item
    }

    override fun pop(): Node {
        //System.out.print("pop -> " + this.peek());
        return super.pop()
    }

    private fun initializeBackButton() {
        backButton = Button("Back")
        backButton!!.setOnAction { event ->
            this.pop()
            MainScene.mainScene.borderPane.center = super.peek()
            if (this.size == 1) {
                MainScene.mainScene.windowScene!!.win.titlebarButtonBox.children.remove(backButton)
                backButton = null
            }
        }
    }
}
